src/input.js

```diff
@@ -58,35 +64,186 @@ function maybeSetSelection(state, dispatch, selection) {
+      } else if (child.nodeType == 3) {
+        boxes = textRange(child, 0, child.nodeValue.length).getClientRects();
+      } else {
+        continue;
+      }
+      for (let i = 0; i < boxes.length; i++) {
+        const box = boxes[i];
+        if (
+          box.bottom > box.top + 1 &&
+          (dir == 'up'
+            ? coords.top - box.top > (box.bottom - coords.top) * 2
+            : box.bottom - coords.bottom > (coords.bottom - box.top) * 2)
+        )
+          return false;
+      }
+    }
+    return true;
+  });
+}
+
+const maybeRTL = /[\u0590-\u08ac]/;
+
+function endOfTextblockHorizontal(view, state, dir, selection) {
+  const {$head} = selection;
+  if (!$head.parent.isTextblock) return false;
+  const offset = $head.parentOffset,
+    atStart = !offset,
+    atEnd = offset == $head.parent.content.size;
+  const sel = view.root.getSelection();
+  // If the textblock is all LTR, or the browser doesn't support
+  // Selection.modify (Edge), fall back to a primitive approach
+  if (!maybeRTL.test($head.parent.textContent) || !sel.modify)
+    return dir == 'left' || dir == 'backward' ? atStart : atEnd;
+
+  return withFlushedState(view, state, () => {
+    // This is a huge hack, but appears to be the best we can
+    // currently do: use `Selection.modify` to move the selection by
+    // one character, and see if that moves the cursor out of the
+    // textblock (or doesn't move it at all, when at the start/end of
+    // the document).
+    const oldRange = sel.getRangeAt(0),
+      oldNode = sel.focusNode,
+      oldOff = sel.focusOffset;
+    const oldBidiLevel = sel.caretBidiLevel; // Only for Firefox
+    sel.modify('move', dir, 'character');
+    const parentDOM = $head.depth
+      ? view.docView.domAfterPos($head.before())
+      : view.dom;
+    const result =
+      !parentDOM.contains(
+        sel.focusNode.nodeType == 1 ? sel.focusNode : sel.focusNode.parentNode
+      ) ||
+      (oldNode == sel.focusNode && oldOff == sel.focusOffset);
+    // Restore the previous selection
+    sel.removeAllRanges();
+    sel.addRange(oldRange);
+    if (oldBidiLevel != null) sel.caretBidiLevel = oldBidiLevel;
+    return result;
+  });
+}
+
+// Check whether the cursor is at the end of a cell (so that further
+// motion would move out of the cell)
+function atEndOfCellFromSelection(view, axis, dir, sel) {
+  if (sel.toJSON().type !== 'text') return null;
+  const {$head} = sel;
+  if (!$head) return null;
+  for (let d = $head.depth - 1; d >= 0; d--) {
+    const parent = $head.node(d),
+      index = dir < 0 ? $head.index(d) : $head.indexAfter(d);
+    if (index != (dir < 0 ? 0 : parent.childCount)) return null;
+    if (
+      parent.type.spec.tableRole == 'cell' ||
+      parent.type.spec.tableRole == 'header_cell'
+    ) {
+      const cellPos = $head.before(d);
+      const dirStr =
+        axis == 'vert' ? (dir > 0 ? 'down' : 'up') : dir > 0 ? 'right' : 'left';
+
+      return (axis == 'vert'
```

Comment:
document?

Reply:
same as `getIcon`, will add documentation to the parts we changed